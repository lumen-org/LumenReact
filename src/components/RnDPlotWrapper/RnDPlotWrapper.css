.RndPlot-container {
  border: #dbdbdb 3px solid;
  border-radius: 10px;
}

.RndPlot-plot {
  width: 95%;
  height: 95%;
}

.RndPlot-titlebar {
  display: flex;
}
