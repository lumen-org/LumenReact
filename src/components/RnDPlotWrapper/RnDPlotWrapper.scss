@import "src/assets/styles/colors";

.RndPlot-container {
  border: $rnd-container-border 3px solid;
  border-radius: 10px;
  overflow: hidden;
  background-color: $rnd-container-background-color;
}

.RndPlot-plot {
  width: 95%;
  height: 95%;
}

.RndPlot-titlebar {
  display: flex;
  overflow: hidden;
}

.RndPlot-continuous-white-background {
  background-color: white;
}