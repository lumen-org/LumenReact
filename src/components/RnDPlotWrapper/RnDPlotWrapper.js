import React, { Component } from "react";
import PropTypes from "prop-types";
import { Rnd } from "react-rnd";
import CloseButton from "../Button/CloseButton";
import MultiPlot from "../MultiPlot";
import StandardPlot from "../StandardPlot";
import PCIGraph from "../PCIGraph";
import {
  STANDARD_PLOT,
  MULTI_PLOT,
  PCI_PLOT,
  DIFFERENTIAL_MARGINAL_PLOT,
} from "../../constants/plotTypes";
import "./RnDPlotWrapper.scss";

/**
 * RnDPlotWrapper is a wrapper around different kinds of plots such as:
 * Differential Marginal Plot, Graph View of Model, Individual Conditionally Expectation Plot
 */

class RnDPlotWrapper extends Component {
  static propTypes = {
    id: PropTypes.number,
    onActivePlotChange: PropTypes.func,
    onPlotClose: PropTypes.func,
    activePlotId: PropTypes.number,
    zIndex: PropTypes.number,
    plotType: PropTypes.string,
  };

  state = {
    plotWindowsWidth: 500,
    plotWindowsHeight: 500,
    plotWindowsPosX: 50,
    plotWindowsPosY: 50,
  };

  setNewPos = (dragIndex) => {
    this.setState({
      plotWindowsPosX: dragIndex.x,
      plotWindowsPosY: dragIndex.y,
    });
  };

  onDragStop = (event, dragIndex) => {
    const { id, onActivePlotChange, activePlotId } = this.props;
    if (id !== activePlotId) {
      onActivePlotChange(id);
    }
    this.setNewPos(dragIndex);
  };

  onResizeStop = (event, direction, ref, delta, position) => {
    this.setState({
      plotWindowsWidth: ref.style.width,
      plotWindowsHeight: ref.style.height,
      ...position,
    });
  };

  handleClose = () => {
    const { onPlotClose, id } = this.props;
    onPlotClose(id);
  };

  render() {
    const { zIndex, id, plotType } = this.props;
    const {
      plotWindowsHeight,
      plotWindowsWidth,
      plotWindowsPosX,
      plotWindowsPosY,
    } = this.state;

    return (
      <Rnd
        size={{ width: plotWindowsWidth, height: plotWindowsHeight }}
        position={{ x: plotWindowsPosX, y: plotWindowsPosY }}
        style={{ zIndex: zIndex }}
        onDragStop={this.onDragStop}
        onResizeStop={this.onResizeStop}
        className="RndPlot-container"
      >
        <div className="RndPlot-titlebar">
          <CloseButton handleClose={this.handleClose} />
        </div>
        {plotType === STANDARD_PLOT ? (
          <StandardPlot id={id} />
        ) : plotType === MULTI_PLOT ? (
          <MultiPlot id={id} />
        ) : plotType === PCI_PLOT ? (
          <PCIGraph id={id} />
        ) : null}
      </Rnd>
    );
  }
}

export default RnDPlotWrapper;
