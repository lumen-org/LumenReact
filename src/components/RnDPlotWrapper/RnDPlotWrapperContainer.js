import React from "react";
import { connect } from "react-redux";
import { nextActiveId } from "../../utils/plotData";
import { changeActivePlot, deletePlot } from "../../states/plots/actions";
import { updateActiveModel } from "../../states/app/actions";
import PropTypes from "prop-types";
import RnDPlotWrapper from "./RnDPlotWrapper";
import {
  changeActiveVisualization,
  deleteVisualization,
} from "../../states/visualizations/actions";
import {
  selectActiveSpecificationId,
  selectActivePlotType,
} from "../../states/visualizations/selector";
import { getPlotTypeById } from "../../states/plots/selector";

class RnDPlotWrapperContainer extends React.Component {
  static propTypes = {
    id: PropTypes.number, // refractor needed, change id to visId
    activePlotId: PropTypes.number,
    activePlotType: PropTypes.string,
    zIndex: PropTypes.number,
  };

  onPlotClose = (id) => {
    const {
      deletePlot,
      plots,
      changeActivePlot,
      changeActiveVisualization,
      deleteVisualization,
      activePlotType,
    } = this.props;
    const nextId = plots.allIds[0];
    const getVisualizationId = (plotId) => plots.byId[plotId].visualizationId;
    changeActiveVisualization(getVisualizationId(nextId));
    // deleteSpecification(getVisualizationId(id));
    // deleteModel(id);
    deletePlot(id, activePlotType);
    deleteVisualization(getVisualizationId(id));
  };

  onActivePlotChange = (id) => {
    const {
      changeActivePlot,
      updateActiveModel,
      modelName,
      deleteVisualization,
      changeActiveVisualization,
      visualizationId,
    } = this.props;
    changeActiveVisualization(visualizationId);
    changeActivePlot(id);
    updateActiveModel(modelName);
  };

  componentDidMount() {}

  render() {
    const { activePlotId, id, zIndex } = this.props;
    return (
      <RnDPlotWrapper
        zIndex={zIndex}
        id={id}
        activePlotId={activePlotId}
        onPlotClose={this.onPlotClose}
        onActivePlotChange={this.onActivePlotChange}
      />
    );
  }
}

const mapStateToProps = (state, ownProps) => ({
  plots: state.plots.plots,
  activePlotId: state.plots.activePlotId,
  activeSpecification: selectActiveSpecificationId(state),
  activePlotType: selectActivePlotType(state),
});

const mapDispatchToProps = (dispatch) => {
  return {
    changeActiveVisualization: (newActiveModelId) =>
      dispatch(changeActiveVisualization(newActiveModelId)),
    changeActivePlot: (
      newActivePlotId // this function change the zIndex of plot and bring it to the front
    ) => dispatch(changeActivePlot(newActivePlotId)),
    updateActiveModel: (
      newActiveModel // this function trigger the update of models
    ) => dispatch(updateActiveModel(newActiveModel)),
    deleteVisualization: (id) => dispatch(deleteVisualization(id)),
    deletePlot: (id, activePlotType) =>
      dispatch(deletePlot(id, activePlotType)),
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(RnDPlotWrapperContainer);
