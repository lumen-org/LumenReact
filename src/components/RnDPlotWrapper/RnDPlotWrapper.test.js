import { unmountComponentAtNode } from "react-dom";
import React from "react";
import { act } from "react-dom/test-utils";
import RnDPlotWrapper from "./RnDPlotWrapper";
import TestRenderer from "react-test-renderer";

jest.mock("../../components/Button/CloseButton", () => (props) =>
  <div>CloseButton</div>
);

let spy = null;
beforeEach(() => {
  spy = jest.spyOn(console, 'error').mockImplementation();
});

afterEach(() => {
  // cleanup on exiting
  spy.mockRestore();
});

describe("", () => {
  it("values", () => {

  })
})

describe("RnDPlotWrapper Funcitionality", () => {
  it("renders with or without a name", () => {
    let component = null;
    act(() => {
      let id = 0;
      let onActivePlotChange = jest.fn();
      let onPlotClose = jest.fn();
      let activePlotId = 1;
      let zIndex = 0;
      let plotType = "Standard";
      component = TestRenderer.create(
        <RnDPlotWrapper
          id={id}
          onActivePlotChange={onActivePlotChange}
          onPlotClose={onPlotClose}
          activePlotId={activePlotId}
          zIndex={zIndex}
          plotType={plotType}
        >
          {<ExampleChildren></ExampleChildren>}
        </RnDPlotWrapper>
      );
    });
    expect(component.toJSON()).toMatchSnapshot();
  });
})


class ExampleChildren extends React.Component {
  render() {
    return <div></div>;
  }
}
