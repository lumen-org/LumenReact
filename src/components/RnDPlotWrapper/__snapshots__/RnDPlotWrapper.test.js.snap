// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RnDPlotWrapper Funcitionality renders with or without a name 1`] = `
<div
  className="RndPlot-container react-draggable"
  onClick={[Function]}
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  style={
    Object {
      "boxSizing": "border-box",
      "cursor": "move",
      "display": "inline-block",
      "flexShrink": 0,
      "height": "500px",
      "left": 0,
      "maxHeight": 9007199254740991,
      "maxWidth": 9007199254740991,
      "minHeight": undefined,
      "minWidth": undefined,
      "position": "absolute",
      "top": 0,
      "transform": "translate(50px,50px)",
      "userSelect": "auto",
      "width": "500px",
      "zIndex": 0,
    }
  }
  transform={null}
>
  <div
    className="RndPlot-titlebar"
  >
    <div>
      CloseButton
    </div>
  </div>
  <div
    className="cancel"
  >
    <div />
  </div>
  <span>
    <div
      className=""
      onMouseDown={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "cursor": "row-resize",
          "height": "10px",
          "left": "0px",
          "position": "absolute",
          "top": "-5px",
          "userSelect": "none",
          "width": "100%",
        }
      }
    />
    <div
      className=""
      onMouseDown={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "cursor": "col-resize",
          "height": "100%",
          "position": "absolute",
          "right": "-5px",
          "top": "0px",
          "userSelect": "none",
          "width": "10px",
        }
      }
    />
    <div
      className=""
      onMouseDown={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "bottom": "-5px",
          "cursor": "row-resize",
          "height": "10px",
          "left": "0px",
          "position": "absolute",
          "userSelect": "none",
          "width": "100%",
        }
      }
    />
    <div
      className=""
      onMouseDown={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "cursor": "col-resize",
          "height": "100%",
          "left": "-5px",
          "position": "absolute",
          "top": "0px",
          "userSelect": "none",
          "width": "10px",
        }
      }
    />
    <div
      className=""
      onMouseDown={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "cursor": "ne-resize",
          "height": "20px",
          "position": "absolute",
          "right": "-10px",
          "top": "-10px",
          "userSelect": "none",
          "width": "20px",
        }
      }
    />
    <div
      className=""
      onMouseDown={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "bottom": "-10px",
          "cursor": "se-resize",
          "height": "20px",
          "position": "absolute",
          "right": "-10px",
          "userSelect": "none",
          "width": "20px",
        }
      }
    />
    <div
      className=""
      onMouseDown={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "bottom": "-10px",
          "cursor": "sw-resize",
          "height": "20px",
          "left": "-10px",
          "position": "absolute",
          "userSelect": "none",
          "width": "20px",
        }
      }
    />
    <div
      className=""
      onMouseDown={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "cursor": "nw-resize",
          "height": "20px",
          "left": "-10px",
          "position": "absolute",
          "top": "-10px",
          "userSelect": "none",
          "width": "20px",
        }
      }
    />
  </span>
</div>
`;
